cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -410,7 +410,7 @@ public void removeApplication(final Id.Application appId) throws Exception {
                        new com.google.common.base.Predicate<ProgramRuntimeService.RuntimeInfo>() {
       @Override
       public boolean apply(ProgramRuntimeService.RuntimeInfo runtimeInfo) {
-        return runtimeInfo.getProgramId().toEntityId().getApplication().equals(appId.getId());
+        return runtimeInfo.getProgramId().toEntityId().getParent().equals(appId.toEntityId());
```

Comment:
Versioning support is from 3.6 and we don't plan to pick this to 3.6+ I saw that you have changed this in 3.6  to do  the correct check. Is that not enough ?